### PR TITLE
Add notebook chunk output gallery

### PR DESCRIPTION
**Pull request by jmcphers**
This isn't ready for merge (pending UX tweaks and other fine-tuning), but it's close enough to its final form for code review.

This is primarily a UX change; there are only a couple of server-side changes involved. They are as follows.

1. **Add support for arbitrary chunk output metadata**. The metadata is used to store properties about the outputs such as their class and dimension for display in the gallery. The implementation is an untyped JSON object stored with each output, persisted in the cache as a `.metadata` file, stored in the `.nb.html` file, and hydrated with the rest of the cache data. 

2. **Implement support for hydrating `.rdf` files**. These were not previously stored in the `.nb.html` file. They are now stored there in base64 form, and deserialized back into the cache when it's hydrated from a `.nb.html` file. 

On the UX side, this is a refactor of the `ChunkOutputWidget`, which was formerly responsible for both receiving data and presenting it. There are now two strategies for presenting output (represented by the `ChunkOutputPresenter` interface): the `ChunkOutputStream` lays out content linearly (a straightforward port of the existing code), and the `ChunkOutputGallery` lays it out in gallery pages. The `ChunkOutputWidget` is responsible for selecting the appropriate presentation model, and routing data to the current presentation model. This makes it relatively easy to adjust the choice of presentation model if e.g. we decide later that we're switching to gallery view too aggressively. 

The new `ChunkOutputGallery` presenter contains a sequence of `ChunkOutputPage` objects, which are responsible for supplying their own thumbnails and content. Each type of content has its own output page class.

### CODE REVIEW

**Comment by jjallaire** [comment]
I'll invoke this well known axiom here: https://twitter.com/girayozil/status/306836785739210752

So, LGTM :-)

Seriously though, I scanned the C++ code for any idomatic issues and/or potential sources of abnormal program endings and saw none. I don't have the context to review the rest of the code in any depth.

**Comment by jmcphers** [comment]
Heh, that's so true. :-) Appreciate the quick review! 

@kevinushey might be interested in double-checking the server-side work items as they both touch `.nb.html` generation and cache reconstruction.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkDataWidget.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+         Element parent) /*-{
+      var pagedTable = $doc.createElement("div");
+      pagedTable.setAttribute("data-pagedtable", "");
+      parent.appendChild(pagedTable);
+
+      var pagedTableSource = $doc.createElement("script");
+      pagedTableSource.setAttribute("data-pagedtable-source", "");
+      pagedTableSource.setAttribute("type", "application/json");
+      pagedTableSource.appendChild($doc.createTextNode(JSON.stringify(data)))
+      pagedTable.appendChild(pagedTableSource);
```
Comment (javierluraschi) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkDataWidget.java]: @jmcphers made a couple changes to this in the old file, see: https://github.com/rstudio/rstudio/commit/414173c5e3c419c6c3b118e5c33e49703e5eafc8

Comment (javierluraschi) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkDataWidget.java]: Or you can just merge this and I'll re-apply them, up to you.

**Comment by kevinushey** [comment]
Did a quick review of the server-side items and all looks well.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkDataWidget.java (hunk 1 above)
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/ChunkDataWidget.java]: Thanks for the heads up! Here's a factored version that uses the new editor colors class + the new data widget: https://github.com/rstudio/rstudio/commit/6bc6ccc8e845761064c9dabead76df01f9660c91

**Comment by kevinushey** [comment]
🎉 !

**Pull request merged**